Proyecto 9. Beta Bank

Paso 1. importar los datasets utilizados.

Primero importaremos las librerías correspondiente. Luego de esto leemos el archivo como valores separados por comas (es CSV, de acuerdo a la terminación del archivo).

In [1]:
#Librerías
import pandas as pd
import numpy as np

from sklearn.preprocessing import OrdinalEncoder, StandardScaler
from sklearn.utils import shuffle
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix, accuracy_score, roc_curve, f1_score, precision_recall_curve, roc_auc_score

In [2]:
##Leemos el archivo
churn = pd.read_csv('/datasets/Churn.csv') 

Paso 2. Observar los datos que contiene y los tipos de datos.

Se imprimen las primeras filas para verificar cual es el delimitador del archivo, como están los nombres de las columnas, entre otras cosas. Se lee la información general usando el método "info()", observando la presencia de datos nulos y el tipo de datos que tenemos en los data frames.

In [3]:
#Mostramos la información del Data frame de us_behave
display(churn)
churn.info()
display(churn.describe())

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0        2.0       0.00              1          1               1   
1        1.0   83807.86              1          0               1   
2        8.0  159660.80              3          1               0   
3        1.0       0.00              2          0               0   
4        2.0  125510.82              1          1               1   
...      ...        ...            ...        ...             ...   
9995     5.0       0.00              2          1               0   
9996    10.0   57369.61              1          1               1   
9997     7.0       0.00              1          0               1   
9998     3.0   75075.31              2          1               0   
9999     NaN  130142.79              1          1               0   

      EstimatedSalary  Exited  
0           101348.88       1  
1           112542.58       0  
2           113931.57       1  
3            93826.63       0  
4            79084.10       0  
...               ...     ...  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0  

[10000 rows x 14 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


RowNumber    CustomerId   CreditScore           Age       Tenure  \
count  10000.00000  1.000000e+04  10000.000000  10000.000000  9091.000000   
mean    5000.50000  1.569094e+07    650.528800     38.921800     4.997690   
std     2886.89568  7.193619e+04     96.653299     10.487806     2.894723   
min        1.00000  1.556570e+07    350.000000     18.000000     0.000000   
25%     2500.75000  1.562853e+07    584.000000     32.000000     2.000000   
50%     5000.50000  1.569074e+07    652.000000     37.000000     5.000000   
75%     7500.25000  1.575323e+07    718.000000     44.000000     7.000000   
max    10000.00000  1.581569e+07    850.000000     92.000000    10.000000   

             Balance  NumOfProducts    HasCrCard  IsActiveMember  \
count   10000.000000   10000.000000  10000.00000    10000.000000   
mean    76485.889288       1.530200      0.70550        0.515100   
std     62397.405202       0.581654      0.45584        0.499797   
min         0.000000       1.000000      0.00000        0.000000   
25%         0.000000       1.000000      0.00000        0.000000   
50%     97198.540000       1.000000      1.00000        1.000000   
75%    127644.240000       2.000000      1.00000        1.000000   
max    250898.090000       4.000000      1.00000        1.000000   

       EstimatedSalary        Exited  
count     10000.000000  10000.000000  
mean     100090.239881      0.203700  
std       57510.492818      0.402769  
min          11.580000      0.000000  
25%       51002.110000      0.000000  
50%      100193.915000      0.000000  
75%      149388.247500      0.000000  
max      199992.480000      1.000000

Tenemos 12 características. Son las siguientes, en su respectivo orden: indice de la cadena de datos, identificador del cliente, apellido, puntaje crediticio, país de residencia, género del cliente, edad, périodo en años de maduración del depostio a plazo fijo de un cliente, saldo de la cuenta, número de productos bancarios usados por el cliente, si tiene (1=si) o no (0=no) una tarjeta de credito, actividad del cliente (1=si, 0=no) y su salario estimado. 
Entre las caractersiticas cabe destacar que Tenure tiene datos nulos, haciendo que tenga datos de tipo float, cuando podrían ser de tipo entero. 

Tenemos un único objetivo: Exited. Este muestra si un cliente se ha ido (1=si) o no (0=no)o)

Paso 3. Preprocesamiento de los datos.

3.1 Verificación de que no tenemos datos duplicados y observación sobre los datos nulos de tenure. 

In [4]:
display(churn.duplicated().sum())

0

In [5]:
display(churn[churn["Tenure"].isna()])

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
30           31    15589475    Azikiwe          591     Spain  Female   39   
48           49    15766205        Yin          550   Germany    Male   38   
51           52    15768193  Trevisani          585   Germany    Male   36   
53           54    15702298   Parkhill          655   Germany    Male   41   
60           61    15651280     Hunter          742   Germany    Male   35   
...         ...         ...        ...          ...       ...     ...  ...   
9944       9945    15703923    Cameron          744   Germany    Male   41   
9956       9957    15707861      Nucci          520    France  Female   46   
9964       9965    15642785    Douglas          479    France    Male   34   
9985       9986    15586914     Nepean          659    France    Male   36   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
30       NaN       0.00              3          1               0   
48       NaN  103391.38              1          0               1   
51       NaN  146050.97              2          0               0   
53       NaN  125561.97              1          0               0   
60       NaN  136857.00              1          0               0   
...      ...        ...            ...        ...             ...   
9944     NaN  190409.34              2          1               1   
9956     NaN   85216.61              1          1               0   
9964     NaN  117593.48              2          0               0   
9985     NaN  123841.49              2          1               0   
9999     NaN  130142.79              1          1               0   

      EstimatedSalary  Exited  
30          140469.38       1  
48           90878.13       0  
51           86424.57       0  
53          164040.94       1  
60           84509.57       0  
...               ...     ...  
9944        138361.48       0  
9956        117369.52       1  
9964        113308.29       0  
9985         96833.00       0  
9999         38190.78       0  

[909 rows x 14 columns]

No se observan datos duplicados en el Dataframe, además de que los datos nulos de Tenure no parecen tener un patrón con otros datos. Como se observó nteriormente, la mediana de los datos de tenure es 5.0. Se usará este dato para llenar los datos nulos.

In [6]:
churn["Tenure"]= churn["Tenure"].fillna(churn["Tenure"].median())
churn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


3.2 Eliminación de columnas que podrían afectar negativamente a nuestro modelo.

Se eliminaran los datos que no parecen ser relevantes para nuestro modelo, como es el caso de las columnas RowNumber, Surname y CustomerId.

In [7]:
churn=churn.drop(columns=['RowNumber', "Surname", "CustomerId"])
display(churn)

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0             619    France  Female   42     2.0       0.00              1   
1             608     Spain  Female   41     1.0   83807.86              1   
2             502    France  Female   42     8.0  159660.80              3   
3             699    France  Female   39     1.0       0.00              2   
4             850     Spain  Female   43     2.0  125510.82              1   
...           ...       ...     ...  ...     ...        ...            ...   
9995          771    France    Male   39     5.0       0.00              2   
9996          516    France    Male   35    10.0   57369.61              1   
9997          709    France  Female   36     7.0       0.00              1   
9998          772   Germany    Male   42     3.0   75075.31              2   
9999          792    France  Female   28     5.0  130142.79              1   

      HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0             1               1        101348.88       1  
1             0               1        112542.58       0  
2             1               0        113931.57       1  
3             0               0         93826.63       0  
4             1               1         79084.10       0  
...         ...             ...              ...     ...  
9995          1               0         96270.64       0  
9996          1               1        101699.77       0  
9997          0               1         42085.58       1  
9998          1               0         92888.52       1  
9999          1               0         38190.78       0  

[10000 rows x 11 columns]

3.3 Estandarización de los datos numéricos no binarios.

Se estandarizarán los datos de las columnas numericas no binarias

In [8]:
# Identificar columnas numéricas
columnas_numericas = churn.select_dtypes(include=['number']).columns

# Identificar columnas binarias (aquellas con exactamente 2 valores únicos)
columnas_binarias = [col for col in columnas_numericas if churn[col].nunique() == 2]

# Identificar columnas numéricas no binarias
columnas_no_binarias = [col for col in columnas_numericas if col not in columnas_binarias]

# Inicializar el escalador estándar
scaler = StandardScaler()

# Estandarizar solo las columnas numéricas no binarias
churn[columnas_no_binarias] = scaler.fit_transform(churn[columnas_no_binarias])
display(churn)

CreditScore Geography  Gender       Age    Tenure   Balance  \
0       -0.326221    France  Female  0.293517 -1.086246 -1.225848   
1       -0.440036     Spain  Female  0.198164 -1.448581  0.117350   
2       -1.536794    France  Female  0.293517  1.087768  1.333053   
3        0.501521    France  Female  0.007457 -1.448581 -1.225848   
4        2.063884     Spain  Female  0.388871 -1.086246  0.785728   
...           ...       ...     ...       ...       ...       ...   
9995     1.246488    France    Male  0.007457  0.000761 -1.225848   
9996    -1.391939    France    Male -0.373958  1.812439 -0.306379   
9997     0.604988    France  Female -0.278604  0.725432 -1.225848   
9998     1.256835   Germany    Male  0.293517 -0.723910 -0.022608   
9999     1.463771    France  Female -1.041433  0.000761  0.859965   

      NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0         -0.911583          1               1         0.021886       1  
1         -0.911583          0               1         0.216534       0  
2          2.527057          1               0         0.240687       1  
3          0.807737          0               0        -0.108918       0  
4         -0.911583          1               1        -0.365276       0  
...             ...        ...             ...              ...     ...  
9995       0.807737          1               0        -0.066419       0  
9996      -0.911583          1               1         0.027988       0  
9997      -0.911583          0               1        -1.008643       1  
9998       0.807737          1               0        -0.125231       1  
9999      -0.911583          1               0        -1.076370       0  

[10000 rows x 11 columns]

Paso 3.2 Codificación ordinal de nuestro Dataframe. 

Se hace la codificación ordinal usando unicamente las columnas con datos de tipo objeto.

In [9]:
col_cat = churn.select_dtypes(include=['object']).columns
# Se crea una instancia de esta clase.
encoder =OrdinalEncoder()
# Se transforman los datos, se obtiene la lista de características categóricas y se agregan los nombres de las columnas
churn[col_cat]= pd.DataFrame(encoder.fit_transform(churn[col_cat]))
display(churn)

CreditScore  Geography  Gender       Age    Tenure   Balance  \
0       -0.326221        0.0     0.0  0.293517 -1.086246 -1.225848   
1       -0.440036        2.0     0.0  0.198164 -1.448581  0.117350   
2       -1.536794        0.0     0.0  0.293517  1.087768  1.333053   
3        0.501521        0.0     0.0  0.007457 -1.448581 -1.225848   
4        2.063884        2.0     0.0  0.388871 -1.086246  0.785728   
...           ...        ...     ...       ...       ...       ...   
9995     1.246488        0.0     1.0  0.007457  0.000761 -1.225848   
9996    -1.391939        0.0     1.0 -0.373958  1.812439 -0.306379   
9997     0.604988        0.0     0.0 -0.278604  0.725432 -1.225848   
9998     1.256835        1.0     1.0  0.293517 -0.723910 -0.022608   
9999     1.463771        0.0     0.0 -1.041433  0.000761  0.859965   

      NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0         -0.911583          1               1         0.021886       1  
1         -0.911583          0               1         0.216534       0  
2          2.527057          1               0         0.240687       1  
3          0.807737          0               0        -0.108918       0  
4         -0.911583          1               1        -0.365276       0  
...             ...        ...             ...              ...     ...  
9995       0.807737          1               0        -0.066419       0  
9996      -0.911583          1               1         0.027988       0  
9997      -0.911583          0               1        -1.008643       1  
9998       0.807737          1               0        -0.125231       1  
9999      -0.911583          1               0        -1.076370       0  

[10000 rows x 11 columns]

Se obtiene las llaves de nuestra codificación ordinal. En la columna Geography tenemos que France es 0, Germany es 1 y Spain es 2. En la columna de Gender 0 es Female y 1 es Male.

In [10]:
for idx, categories in enumerate(encoder.categories_):
    print(f"Característica {idx}:")
    for category, number in zip(categories, range(len(categories))):
        print(f"  {category} -> {number}")

Característica 0:
  France -> 0
  Germany -> 1
  Spain -> 2
Característica 1:
  Female -> 0
  Male -> 1


Paso 4. Equilibrio de clases. Entrenar el modelo sin tener en cuenta el desequilibrio.

In [11]:
frecuencia_valores = churn['Exited'].value_counts()
print(frecuencia_valores)

0    7963
1    2037
Name: Exited, dtype: int64


Observamos que tenemos alrededdor de 3.9091801669 veces más valores negativos (0, que no se han ido) que valores positivos (1, que se han ido). De manera total tenemos 7963 valores negativos y 2037 valores positivos.

Paso 4.1 Segmentar los datos fuente en tres conjuntos: entrenamiento (60%), validación (20%) y prueba (20%).

In [12]:
# Dividir los datos en características y objetivo
features = churn.drop(columns=['Exited'])
target = churn['Exited']

In [13]:
# Dividimos los datos en dos partes: 80% para entrenamiento+validación y 20% para prueba
features_train_val, features_test, target_train_val, target_test = train_test_split(features, target, test_size=0.2, random_state=12345)

# Dividimos el conjunto de entrenamiento+validación en dos partes: 
#75% (60% del total original) para entrenamiento y 25% (20% del total original) para validación
features_train, features_val, target_train, target_val = train_test_split(features_train_val, target_train_val, test_size=0.25, random_state=12345)

Paso 4.2 Modelo de Bosque aleatorio.

Para este modelo determinamos best_score igual a 0 y el número de estimadores (best_est) en 0 árboles de nuestro bosque. De esta manera se itera por los árboles de nuestro rango del hiperparámetro. El modelo se configura por el número de árboles, con el estado random de 12345. Se entrena el modelo con el conjunto de entrenamiento. Se calcula la puntuación de exactitud usando el conjunto de validación. De está forma obtenemos el mejor modelo y su exactitud, con el número de estimadores.

In [14]:
best_score = 0
best_est = 0

for est in range(1, 101): 
    model = RandomForestClassifier(random_state=12345, n_estimators=est)
    model.fit(features_train, target_train)
    score = model.score(features_val, target_val)
    if score > best_score:
        best_score = score
        best_est = est

print("La exactitud del mejor modelo usando el conjunto de validación (n_estimators = {}): {}".format(best_est, best_score))

La exactitud del mejor modelo usando el conjunto de validación (n_estimators = 51): 0.864


La exactitud del mejor modelo usando el conjunto de validación, con 51 estimadores es de 0.864. Esta puntuación de exacttitud se obtuvo sin verificar el equilibrio de clases. 


4.3 Matriz de confusión.

In [15]:
predicted_val = model.predict(features_val)
print(confusion_matrix(target_val, predicted_val))

[[1550   59]
 [ 220  171]]


Observamos que tenemos 1550 verdaderos negativos (VN), 171 verdaderos positivos (VP), 220 falsos negativos (FN) y 59 falsos positivos (FP). Nuestro modelo parece tener una mayor cantidad de negativos (0) que positivos (1).

4.4 Valor de F1.

In [16]:
print(f1_score(target_val, predicted_val))

0.5507246376811594


El valor F1 es una métrica de evaluación que es la media armónica de las métricas de sensibilidad  y precisión Sus valores van de 0 a 1, siendo 1 donde hay una mayor sensibilidad y precisión para la predicción de la clase positiva. En este caso, vemos que nuestro valor F1, con este modelo sin equilibrar es mediocre.

Paso 5. Mejorar la calidad del modelo

5.1 Modelo balanceado, ajuste del peso de clase

Usamos el modelo de bosque aleatorio, con el peso de clase balanceado.

In [17]:
best_score = 0
best_est = 0

for est in range(1, 101): 
    model_eq = RandomForestClassifier(random_state=12345, n_estimators=est, class_weight='balanced')
    model_eq.fit(features_train, target_train)
    score_eq = model_eq.score(features_val, target_val)
    if score_eq > best_score:
        best_score = score_eq
        best_est = est

print("La exactitud del mejor modelo usando el conjunto de validación (n_estimators = {}): {}".format(best_est, best_score))

La exactitud del mejor modelo usando el conjunto de validación (n_estimators = 46): 0.865


In [18]:
predicted_val_eq = model_eq.predict(features_val)
#Matriz de confusión
print("Obtuvimos la matriz de confusión con los siguientes parámetros:", confusion_matrix(target_val, predicted_val_eq))
print()
#Valor de F1
print("El valor de F1 que obtuvimos es de", f1_score(target_val, predicted_val_eq))

Obtuvimos la matriz de confusión con los siguientes parámetros: [[1564   45]
 [ 227  164]]

El valor de F1 que obtuvimos es de 0.5466666666666666


Los resultados no fueron mejores a comparación del modelo no balanceado. A continuación intentaremos el sobremuestreo.

5.2 Sobremuestreo

Primero vemos las dimensiones de las características y objetivos de las respuestas 1 y 0, respectivamente.

In [19]:
features_zeros=features_train[target_train == 0]
features_ones=features_train[target_train == 1]
target_zeros=target_train[target_train == 0]
target_ones=target_train[target_train == 1]

print(features_zeros.shape)
print(features_ones.shape)
print(target_zeros.shape)
print(target_ones.shape)

(4781, 10)
(1219, 10)
(4781,)
(1219,)


Definimos la función upsample():

In [20]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]
    features_upsampled = pd.concat([features_zeros] +
                                   [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] +
                                 [target_ones] * repeat)

    return features_upsampled, target_upsampled

Hay 3.9220672683 veces más respuestas 0 que 1. Multiplicaremos las respuestas positivas(1) 4 veces. Barajeamos los datos para mezclar las observaciones de forma aleatoria. Observamos las dimensiones de nuestras caracteristicas y objetivos.

In [21]:
features_upsampled, target_upsampled = upsample(features_train, target_train, 4)

features_upsampled, target_upsampled = shuffle(features_upsampled, target_upsampled, random_state=12345)

print(features_upsampled.shape)
print(target_upsampled.shape)

(9657, 10)
(9657,)


Hacemos la prueba con el modelo de bosque aleatorio, balanceado y no balanceado

No balanceado

In [ ]:
best_score = 0
best_est = 0

for est in range(1, 101): 
    model_up_no_bal = RandomForestClassifier(random_state=12345, n_estimators=est)
    model_up_no_bal.fit(features_upsampled, target_upsampled)
    score_eq = model_up_no_bal.score(features_val, target_val)
    if score_eq > best_score:
        best_score = score_eq
        best_est = est

print("La exactitud del mejor modelo usando el conjunto de validación (n_estimators = {}): {}".format(best_est, best_score))

Valor F1

In [ ]:
predicted_up_no_bal= model_up_no_bal.predict(features_val)
#Valor de F1
print("El valor de F1 que obtuvimos es de", f1_score(target_val, predicted_up_no_bal))

Balanceado

In [ ]:
best_score = 0
best_est = 0

for est in range(1, 101): 
    model_up_bal= RandomForestClassifier(random_state=12345, n_estimators=est, class_weight='balanced')
    model_up_bal.fit(features_upsampled, target_upsampled)
    score_eq = model_up_bal.score(features_val, target_val)
    if score_eq > best_score:
        best_score = score_eq
        best_est = est

display("La exactitud del mejor modelo usando el conjunto de validación (n_estimators = {}): {}".format(best_est, best_score))

Valor F1

In [ ]:
predicted_up_bal= model_up_bal.predict(features_val)
#Valor de F1
display("El valor de F1 que obtuvimos es de", f1_score(target_val, predicted_up_bal))

Debido al sobremuestreo obtuvimos mejores valores de F1 usando el modelo de bosque aleatorio, balanceado o no. El balanceado tiene un valor F1 y exactitud ligeramente superior a comparación del no balanceado. 

5.3 Ajuste del umbral de clasificación, curva ROC

Obtenemos las probabilidades predichas de la clase positiva (1).

In [ ]:
probabilities_valid = model_up_bal.predict_proba(features_val)
probabilities_one_valid = probabilities_valid[:, 1]

Calculamos el umbral óptimo y el valor F1 de mayor puntaje.

In [ ]:
# Calculamos la curva ROC
fpr, tpr, thresholds = roc_curve(target_val, probabilities_one_valid)

# Calculamos el F1-score para diferentes umbrales
f1_scores = [f1_score(target_val, probabilities_one_valid >= t) for t in thresholds]

# Encontramos el umbral que maximiza el F1-score
optimal_threshold = thresholds[np.argmax(f1_scores)]
max_f1_score = max(f1_scores)

print(f"Umbral óptimo: {optimal_threshold:.2f}")
print(f"F1-score máximo: {max_f1_score:.2f}")

# Hacemos predicciones con el umbral óptimo
y_pred_optimal = (probabilities_one_valid >= optimal_threshold).astype(int)

# Calculamos el F1-score y otras métricas
final_f1_score = f1_score(target_val, y_pred_optimal)
print(f"F1-score con umbral óptimo: {final_f1_score:.2f}")

De esta forma obtenemos que el umbral óptimo es de 0.46 a comparación del umbral inicial de 0.5 y de esta forma obtenemos un valor de F1 redondeado de 0.59. Este valor sugiere que el modelo tiene un rendimiento moderado en términos de equilibrio entre precisión y recall. Ni la precisión ni el recall son especialmente altos.

5.4 Métrica AUC-ROC

AUC-ROC equivale al área bajo la curva ROC (Receiver Operating Characteristic). Los valores están en el rango de 0 a 1. Para un modelo aleatorio el valor es de 0.5.

In [ ]:
auc_roc = roc_auc_score(target_val, probabilities_one_valid)
print(f"El valor de AUC-ROC es de: {auc_roc:.2f}")

Obtuvimos un valor redondeado de 0.85 de la métrica AUC-ROC. El modelo parece tener una buena capacidad para distinguir entre las clases positiva y negativa. Está funcionando mejor que una predicción aleatoria.

Paso 6. Prueba del modelo.

Con estos valores de AUC-ROC (0.85) y con el valor máximo de F1 (0.59), usando el modelo de bosque aleatorio con una exactitud de 0.8545 con 98 estimadores, tenemos un modelo para predecir si un cliente dejará o no el banco.

In [ ]:
model_test= RandomForestClassifier(random_state=12345, n_estimators=98, class_weight='balanced')
model_test.fit(features_upsampled, target_upsampled)
score_test = model_test.score(features_test, target_test)
print("Puntuación de exactitud usando el conjunto de prueba, con 98 estimadores, balanceado y con sobremuestreo es de", score_test)

Nuestro modelo muestra indicios de sobreajuste. La exactitud obtenida usando el conjunto de prueba es de 0.842. Esto es ligeramente menor a comparación de la puntuación obtenida usando el conjunto de validación: 0.8545. Esto es comprensible. Estamos usando una quinta parte de la cantidad de datos iniciales.